src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs

```diff
@@ -581,39 +614,48 @@ internal static string UnmangleMetadataNameForArity(string emittedTypeName, int
 
             if (dots == 0)
             {
-                return name == SystemString ? s_splitQualifiedNameSystem : ImmutableArray.Create(name);
+                return nameMemory.Span.SequenceEqual(SystemString.AsSpan()) ? splitSystemString : ImmutableArray.Create(convert(nameMemory));
             }
 
-            var result = ArrayBuilder<string>.GetInstance(dots + 1);
+            var result = ArrayBuilder<T>.GetInstance(dots + 1);
 
             int start = 0;
             for (int i = 0; dots > 0; i++)
             {
-                if (name[i] == DotDelimiter)
+                if (nameSpan[i] == DotDelimiter)
                 {
                     int len = i - start;
-                    if (len == 6 && start == 0 && name.StartsWith(SystemString, StringComparison.Ordinal))
+                    if (len == 6 && start == 0 && nameSpan.StartsWith(SystemString.AsSpan(), StringComparison.Ordinal))
                     {
-                        result.Add(SystemString);
+                        result.Add(convert(SystemString.AsMemory()));
                     }
                     else
                     {
-                        result.Add(name.Substring(start, len));
+                        result.Add(convert(nameMemory.Slice(start, len)));
                     }
 
                     dots--;
                     start = i + 1;
                 }
             }
 
-            result.Add(name.Substring(start));
+            result.Add(convert(nameMemory[start..]));
 
             return result.ToImmutableAndFree();
         }
 
         internal static string SplitQualifiedName(
             string pstrName,
             out string qualifier)
+        {
+            ReadOnlyMemory<char> nameMemory = SplitQualifiedName(pstrName, out ReadOnlyMemory<char> qualifierMemory);
+            qualifier = qualifierMemory.ToString();
+            return nameMemory.ToString();
```

Comment:
>nameMemory.ToString() [](http://example.com/codeflow?start=19&length=21)

Probably should reuse the original string when possible.

Reply:
this has those semantics in the caser that we the memory is a wrapper around a string instnace.